### PR TITLE
Prepare for removal of `.spec.dns.{ingressDomain,provider.domains,provider.zones}` fields in the Seed API

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area dev-productivity
/kind cleanup

**What this PR does / why we need it**:
With https://github.com/gardener/gardener/pull/3131, the `.spec.dns.ingressDomain` field was deprecated more than two years ago in favor of `.spec.dns.provider` and `.spec.ingress`. 
This PR removes the usages of this field in the controllers.
The field will be removed from the API in a later release.

**Which issue(s) this PR fixes**:
Part of #7519

**Special notes for your reviewer**:

**Release note**:

```breaking operator
All the functionality related to the deprecated field `spec.dns.ingressDomain` in the Seed has been removed. Subsequently, `.spec.dns.ingressDomain` will be dropped from the Seed API in a later release of Gardener. Before upgrading to this Gardener version, make sure to switch all seeds to `.spec.ingress` and `.spec.dns.provider`. Please find more details here, https://github.com/gardener/gardener/blob/master/docs/deployment/deploy_gardenlet_manually.md#kubernetes-cluster-that-should-be-registered-as-a-seed-cluster
```
```breaking operator
The fields `.spec.dns.provider.domains` and `.spec.dns.provider.zones` in the Seed are deprecated, and all the functionality related to the deprecated fields has been removed. Subsequently these fields will be dropped from the Seed API in a later release of Gardener. Before upgrading to this Gardener version, make sure to switch all seeds not to use these fields.
```

### CODE REVIEW

File: pkg/controllermanager/controller/managedseedset/replica_test.go [hunk 1]
```diff
@@ -373,11 +367,7 @@ var _ = Describe("Replica", func() {
 									Object: &gardenletv1alpha1.GardenletConfiguration{
 										SeedConfig: &gardenletv1alpha1.SeedConfig{
 											SeedTemplate: gardencorev1beta1.SeedTemplate{
-												Spec: gardencorev1beta1.SeedSpec{
-													DNS: gardencorev1beta1.SeedDNS{
-														IngressDomain: pointer.String("ingress." + replicaName + ".example.com"),
```
Comment (timuthy) [inline pkg/controllermanager/controller/managedseedset/replica_test.go:378]: Why was this expectation not replaced by a proper `.spec.ingress.domain` expectation? Earlier, this test expected a calculated `IngressDomain` field which does not happen anymore.

**Review by timuthy — COMMENTED** [review]
Thanks for the PR 🙂 

In a joint review with @oliver-goetz and @darmiel we wondered why the API deprecation  changes are not already part of this PR. We think it might be better to ship both changes in one release, i.e. have them in one PR, mainly for usability reasons.

Does it make sense to include the API field deprecation changes in this PR, as well?
- Add deprecation note to API field.
- Add warning that field shall not be used any more.
- Drop all usages of this field in the code. Today there are still tests that use this field.

**Comment by shafeeqes** [comment]
> Does it make sense to include the API field deprecation changes in this PR, as well?
> 
> * Add deprecation note to API field.
> * Add validation that field must not be used any more.
> * Adapt existing validation.
> * Drop all usages of this field in the code. Today there are still tests that use this field.

This is the approach we follow after #7314, because if the `gardener-apiserver` is upgraded to a version in which this field is deprecated, the API will no longer return this field, hence the controllers which are still using this field will error. That's why we follow this two step approach. Ref: https://gardener-cloud.slack.com/archives/C045DSWJZB9/p1673858499729889

**Comment by timuthy** [comment]
We don't want the field to be removed in this PR but still possible API changes to happen like adding a clear deprecation message as well as printing a warning that this field shall not be used anymore.

**Comment by shafeeqes** [comment]
> We don't want the field to be removed in this PR but still possible API changes to happen like adding a clear deprecation message as well as printing a warning that this field shall not be used anymore.

Ah, right. Okay. Will make the changes.

**Comment by timuthy** [comment]
Sounds good, thank you. Please also try to adjust all existing tests that use this deprecated field to not produce any more usages in the meantime through, .e.g. copy-paste actions.

File: pkg/controllermanager/controller/managedseedset/replica_test.go (hunk 1 above)
Comment (shafeeqes) [reply pkg/controllermanager/controller/managedseedset/replica_test.go:378]: Nice catch, adapted.

File: example/50-seed.yaml [hunk 2]
```diff
@@ -27,29 +27,28 @@ spec:
...
+      #   include:
+      #   - my-custom-domain.com
+      #   - my-other-custom-domain.com
+      #   exclude:
+      #   - yet-another-custom-domain.com
+      # zones:
+      #   include:
+      #   - zone-id-1
+      #   exclude:
+      #   - zone-id-2
```
Comment (rfranzke) [inline example/50-seed.yaml]: Drop this and also deprecate it in the API and remove it similarly with the next step. These fields are not used since we have introduced the `DNSRecord` extensions API years back.

File: pkg/registry/core/seed/strategy.go [hunk 3]
```diff
@@ -71,7 +72,10 @@ func (s Strategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object)
 // Validate validates the given object.
 func (Strategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
 	seed := obj.(*core.Seed)
-	return validation.ValidateSeed(seed)
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validation.ValidateSeed(seed)...)
+	allErrs = append(allErrs, validation.ValidateIngressDomain(seed)...)
+	return allErrs
```
Comment (rfranzke) [inline pkg/registry/core/seed/strategy.go]: Please revert this and call your new logic from within `ValidateSeed`

File: test/integration/gardenlet/seed/seed/seed_test.go [hunk 4]
```diff
@@ -188,6 +192,21 @@ var _ = Describe("Seed controller tests", func() {
 		))
 		DeferCleanup(test.WithFeatureGate(gardenletfeatures.FeatureGate, features.HVPA, true))
 
+		By("Create dns provider secret in garden namespace")
```
Comment (rfranzke) [inline test/integration/gardenlet/seed/seed/seed_test.go]:
```suggestion
		By("Create DNS provider secret in garden namespace")
```

File: test/integration/gardenlet/seed/seed/seed_test.go [hunk 5]
```diff
@@ -188,6 +192,21 @@ var _ = Describe("Seed controller tests", func() {
...
+		dnsProviderSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "secret-",
+			Namespace:    testNamespace.Name,
+			Labels: map[string]string{
+				testID: testRunID,
+			},
+		}}
+		Expect(testClient.Create(ctx, dnsProviderSecret)).To(Succeed())
+
+		By("Wait until the manager cache observes the dns provider secret")
```
Comment (rfranzke) [inline test/integration/gardenlet/seed/seed/seed_test.go]:
```suggestion
		By("Wait until the manager cache observes the DNS provider secret")
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: example/50-seed.yaml [hunk 6]
```diff
@@ -27,29 +27,17 @@ spec:
...
+      type: aws-route53
+      secretRef:
+        name: ingress-secret
+        namespace: garden
+  ingress: 
+    domain: ingress.dev.my-seed.example.com
+    controller:
+      kind: nginx
+      providerConfig:
+        <some-optional-config-for-the-nginx-ingress-controller>
```
Comment (rfranzke) [inline example/50-seed.yaml]:
```suggestion
    # providerConfig:
    #   <some-optional-config-for-the-nginx-ingress-controller>
```

File: pkg/apis/core/validation/seed.go [hunk 7]
```diff
@@ -188,44 +188,50 @@ func ValidateSeedSpec(seedSpec *core.SeedSpec, fldPath *field.Path, inTemplate b
...
-		allErrs = append(allErrs, validateDNS1123Subdomain(*seedSpec.DNS.IngressDomain, fldPath.Child("dns", "ingressDomain"))...)
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("dns", "ingressDomain"), seedSpec.DNS.IngressDomain, "this field is deprecated and will be removed in a future version. Use spec.ingress.domain instead"))
+	}
+	if provider := seedSpec.DNS.Provider; provider != nil {
+		if provider.Domains != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("dns", "provider", "domains"), provider.Domains, "this field is deprecated and will be removed in a future version. Don't set this field"))
+		}
+		if provider.Zones != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("dns", "provider", "zones"), provider.Zones, "this field is deprecated and will be removed in a future version. Don't set this field"))
+		}
```
Comment (rfranzke) [inline pkg/apis/core/validation/seed.go:199]: We can't mark this "invalid" - otherwise people can't update properly. You can add client-side warnings, though.

File: plugin/pkg/managedseed/validator/admission.go [hunk 8]
```diff
@@ -284,6 +285,9 @@ func (v *ManagedSeed) admitSeedSpec(spec *gardencore.SeedSpec, shoot *gardencore
 	}
 
 	// Initialize and validate DNS and ingress
+	if spec.Ingress == nil {
+		spec.Ingress = &core.Ingress{}
+	}
```
Comment (rfranzke) [inline plugin/pkg/managedseed/validator/admission.go:290]: With this, the `if` condition in L292 doesn't make sense anymore.

File: test/integration/gardenlet/seed/seed/seed_test.go [hunk 9]
```diff
@@ -188,6 +192,21 @@ var _ = Describe("Seed controller tests", func() {
...
+			Labels: map[string]string{
+				testID: testRunID,
+			},
+		}}
+		Expect(testClient.Create(ctx, dnsProviderSecret)).To(Succeed())
+
+		By("Wait until the manager cache observes the DNS provider secret")
+		Eventually(func() error {
+			return mgrClient.Get(ctx, client.ObjectKeyFromObject(dnsProviderSecret), dnsProviderSecret)
+		}).Should(Succeed())
```
Comment (rfranzke) [inline test/integration/gardenlet/seed/seed/seed_test.go:208]: Add deletion code in a `DeferCleanup`

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/apis/core/validation/seed.go (hunk 7 above)
Comment (shafeeqes) [reply pkg/apis/core/validation/seed.go:199]: Why is that? The release note mentions that these fields should be unset before upgrading to this Gardener version.

Comment (rfranzke) [reply pkg/apis/core/validation/seed.go:199]: OK, then it's fine. Thanks!

**Review by rfranzke — COMMENTED** [review]
/lgtm

File: docs/extensions/dnsrecord.md [hunk 10]
```diff
@@ -30,13 +30,6 @@ The kubeconfig handed out to end-users does contain this *external domain name*,
 As not every end-user has an own domain, it is possible for Gardener administrators to configure so-called *default domains*.
 If configured, shoots that do not specify a domain explicitly get an *external domain name* based on a default domain (unless explicitly stated that this shoot should not get an external domain name (`.spec.dns.provider=unmanaged`).
 
-### Ingress Domain Name (Deprecated)
-
-Gardener allows to deploy a `nginx-ingress-controller` into a shoot cluster (deprecated).
-This controller is exposed via a public load balancer (again, either IP or hostname).
-Gardener creates a wildcard DNS record pointing to this load balancer.
-`Ingress` resources can later use this wildcard DNS record to expose underlying applications.
```
Comment (timuthy) [inline docs/extensions/dnsrecord.md]: Why is this deleted? To my knowledge, a wildcard DNS record is still created if the Nginx Addon is enabled.

In addition, there should be another paragraph for the `seed-ingress` record.

**Review by timuthy — COMMENTED** [review]
Thanks for all the changes. Just one more doubt from my side.

**Review by oliver-goetz — COMMENTED** [review]
Apart @timuthy's comment regarding `dnsrecord.md` lgtm

File: docs/extensions/dnsrecord.md [hunk 11]
```diff
@@ -30,6 +30,17 @@ The kubeconfig handed out to end-users does contain this *external domain name*,
...
+### Ingress Domain Name (Deprecated)
+
+Gardener allows to deploy a `nginx-ingress-controller` into a shoot cluster (deprecated).
+This controller is exposed via a public load balancer (again, either IP or hostname).
+Gardener creates a wildcard DNS record pointing to this load balancer.
+`Ingress` resources can later use this wildcard DNS record to expose underlying applications.
+
+### Seed Ingress
+
+If `.spec.ingress` is configured in the Seed, Gardener deploys the ingress controller mentioned in `.spec.ingress.controller.kind` to the seed cluster. Currently, the only supported kind is "nginx". If the ingress field is set, then `.spec.dns.provider` must also be set. Gardener creates a wildcard DNS record pointing to the load balancer of the ingress controller. The `Ingress` resources of components like Grafana and Prometheus in the `garden` namespace and the shoot namespaces uses this wildcard DNS record to expose their underlying applications.
```
Comment (timuthy) [inline docs/extensions/dnsrecord.md]:
```suggestion
If `.spec.ingress` is configured in the Seed, Gardener deploys the ingress controller mentioned in `.spec.ingress.controller.kind` to the seed cluster. Currently, the only supported kind is "nginx". If the ingress field is set, then `.spec.dns.provider` must also be set. Gardener creates a wildcard DNS record pointing to the load balancer of the ingress controller. The `Ingress` resources of components like Grafana and Prometheus in the `garden` namespace and the shoot namespaces use this wildcard DNS record to expose their underlying applications. 
```

**Review by timuthy — COMMENTED** [review]
/lgtm
/approve
Thank you!
